### PR TITLE
feat: support sequential exchange

### PR DESCRIPTION
**Pull request by chenzl25**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- Try to resolve https://github.com/risingwavelabs/risingwave/issues/14634
- For the exchange of 2 phase (small) limit, we apply a sequential exchange to avoid OOM.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by liurenjie1024** [comment]
It changes the memory usage from `task_count * limit` to `limit`, but it happens on frontend node. Is the frontend node oom?

File: src/batch/src/task/task_execution.rs [hunk 1]
```diff
@@ -196,6 +196,12 @@ pub struct TaskOutput {
     failure: Arc<Mutex<Option<Arc<BatchError>>>>,
 }
 
+impl Drop for TaskOutput {
+    fn drop(&mut self) {
+        println!("drop TaskOutput {:?}", self.output_id);
+    }
+}
```
Comment (chenzl25) [inline src/batch/src/task/task_execution.rs:203]: Previously, for small limit queries, this struct could not be dropped timely.

**Comment by chenzl25** [comment]
> It changes the memory usage from `task_count * limit` to `limit`, but it happens on frontend node. Is the frontend node oom?

CN oom, I think it is related to the early stop of the limit execution.

**Comment by liurenjie1024** [comment]
> > It changes the memory usage from `task_count * limit` to `limit`, but it happens on frontend node. Is the frontend node oom?
> 
> CN oom, I think it is related to the early stop of the limit execution.

Let's take a look at the sysbench.

File: src/batch/src/task/task_execution.rs (hunk 1 above)
Comment (chenzl25) [reply src/batch/src/task/task_execution.rs:203]: Under a constant workload of limit queries. It could cause a OOM.

**Comment by fuyufjh** [comment]
Are we sure this is the cause? Please at least run a sysbench test to verify it before merging.

**Comment by liurenjie1024** [comment]
> Are we sure this is the cause? Please at least run a sysbench test to verify it before merging.

+1

**Comment by BugenZhao** [comment]
We'd better identify the root cause of the regression before applying any other optimizations.

**Comment by chenzl25** [comment]
Previous CN memory: 
QPS of limit select: 500

image

**Comment by chenzl25** [comment]
After this PR:
QPS of limit select: 1500
CN Memory:
image

**Comment by chenzl25** [comment]
It has achieved 300% performance improvement and much more stable memory consumption under this workload.

**Comment by chenzl25** [comment]
The root cause is small limit queries would fire more tasks as needed in the exchange operators, but when the limit query is ended, those fired tasks cannot be `dropped` timely, because it relied on the channel closed and the drop is asynchronous. Under a constantly limit queries workload, it would end up with too many tasks being dropped, which finally causes an OOM.

**Comment by chenzl25** [comment]
This issue reminds me of the backfill prefetch streaming read OOM issue. Both of them share the same behavior.

File: src/batch/src/executor/generic_exchange.rs [hunk 2]
```diff
@@ -164,26 +167,33 @@ impl<CS: 'static + Send + CreateSource, C: BatchTaskContext> Executor
...
-                    )
-                })
-                .collect_vec(),
-        )
-        .boxed();
-
-        while let Some(data_chunk) = stream.next().await {
-            let data_chunk = data_chunk?;
-            yield data_chunk
+        let streams = self
```
Comment (liurenjie1024) [inline src/batch/src/executor/generic_exchange.rs:170]: How about move this part to the `self.sequential` block? It's not easy to understand the difference of collection.

**Review by liurenjie1024 — APPROVED** [review]
LGTM, thanks!

**Comment by chenzl25** [comment]
Sysbench test passed.
https://buildkite.com/risingwave-test/sysbench/builds/667#018d44ba-5ba0-49b7-a472-a952fe9134eb

image

**Review by BugenZhao — COMMENTED** [review]
Can we document what `sequential` is somewhere in the code? It seems to be not very intuitive.

**Pull request merged**